File: 1.array/303.RangeSumQuery/main.go
package main

// https://leetcode.com/problems/range-sum-query-immutable/

type NumArray struct {
	prefixSum []int
}

// precalculate the sum of leftside
// [-2,0,3,-5,2,-1]
// [-2,-2,1,-4,-2,-3]
func Constructor(nums []int) NumArray {
	prefixSum := make([]int, len(nums)+1)
	prefixSum[0] = 0
	for i := 1; i< len(prefixSum); i++ {
		prefixSum[i] = prefixSum[i-1] + nums[i-1]
	}
	return NumArray{
		prefixSum: prefixSum,
	}
}

// sum[left:right] = prefixSum[right] - prefixSum[left]
func (this *NumArray) SumRange(left int, right int) int {
	return this.prefixSum[right+1] - this.prefixSum[left]
}

/**
* Your NumArray object will be instantiated and called as such:
* obj := Constructor(nums);
* param_1 := obj.SumRange(left,right);
 */

File: 1.array/303.RangeSumQuery/main.py
from typing import List

# SC: O(n)
class NumArray:
	# TC: O(n)
    def __init__(self, nums: List[int]):
        n = len(nums)
        prefix = [0] * (n + 1)
        for i in range(n):
            prefix[i+1] = prefix[i] + nums[i]
        self.prefix = prefix
	# TC: O(1)
    def sumRange(self, left: int, right: int) -> int:
        return self.prefix[right + 1] - self.prefix[left]
